include/GenSync/Syncs/BloomFilter.h

```diff
@@ -0,0 +1,231 @@
+ * whether an element is a member of a set. False positives are possible, but
+ * not false negatives. This data structure was developed by Burton Howard Bloom in 1970.
+ *
+ * Created by Anish Sinha on 12/4/23.
+ */
+
+#ifndef GENSYNCLIB_BLOOMFILTER_H
+#define GENSYNCLIB_BLOOMFILTER_H
+
+#include <vector>
+#include <utility>
+#include <string>
+#include <NTL/ZZ.h>
+#include <sstream>
+#include <GenSync/Aux/Auxiliary.h>
+#include <GenSync/Data/DataObject.h>
+
+using std::hash;
+using std::pair;
+using std::string;
+using std::stringstream;
+using std::vector;
+using namespace NTL;
+
+// Shorthand for the hash type
+typedef unsigned long int hash_t;
+
+class BloomFilter
+{
+public:
+
+    /**
+     * Builder pattern for creating BloomFilter objects
+     */
+    class Builder
+    {
+    public:
+        /** Constructor for builder pattern */
+        Builder(){
+            bfSize = 0;
+            numHashes = 0;
+            numExpElems = 0;
+            falsePosProb = 0;
+        }
+
+        /**
+         * Builds a BloomFilter object.
+         * setSize and setNumHashes are meant to be used together.
+         * setNumExpElems and setFalsePosProb are meant to be used together.
+         * If there is a mismatched use of setters, output error message and quit, returnng empty BloomFilter.
+         * @return a BloomFilter object from the build parts that have been set.
+         */    
+        BloomFilter build()
+        {
+            if(bfSize != 0 && numHashes != 0)
+                return BloomFilter(bfSize, numHashes);
+            
+            if(numExpElems != 0 && falsePosProb != 0)
+                return BloomFilter(numExpElems, falsePosProb);
+            
+            Logger::error_and_quit("ERROR: Mismatched combination of setters used in BloomFilter construction!");
+            return BloomFilter();
+        }        
+
+        /**
+         * Sets the size of the BloomFilter, specifically its length in bits.
+         * Meant to be used with setNumHashes setter. Cannot be used with other setters.
+         * Discards all elements inserted before the resize.
+         */
+        Builder& setSize(size_t size)
+        {
+            bfSize = size;
+            return *this;
+        }
+
+        /**
+         * Sets the number of hash functions used by the BloomFilter for each element insertion
+         * Meant to be used with setSize setter. Cannot be used with other setters.
+         * Discards all elements inserted before the hashes reset.
+         */
+        Builder& setNumHashes(int nHash)
+        {
+            numHashes = nHash;
+            return *this;
+        }
+
+        /**
+         * Sets the number of expected elements in the BloomFilter.
+         * Meant to be used with setFalsePosProb setter. Cannot be used with other setters.
+         * Discards all elements inserted before the resize.
+         */
+        Builder& setNumExpElems(size_t numElems)
+        {
+            numExpElems = numElems;

```

Comment:
Maybe here you can check whether numHashes has already been set, and throw an error if so, reminding the user that the two conflict.

Reply:
I addressed your previous comment, and added a check to the final build method, which throws an error if there are conflicting setters. So, I did not add a check to the setters themselves as it is already done in the build method.